homeassistant/helpers/event.py

```diff
@@ -349,41 +353,346 @@ def remove_listener() -> None:
+    exception, the listener will still be registered but will only
+    fire if the template result becomes true without an exception.
+
+    Action arguments
+    ----------------
+    entity_id
+        ID of the entity that triggered the state change.
+    old_state
+        The old state of the entity that changed.
+    new_state
+        New state of the entity that changed.
+
+    Parameters
+    ----------
+    hass
+        Home assistant object.
+    template
+        The template to calculate.
+    action
+        Callable to call with results. See above for arguments.
+    variables
+        Variables to pass to the template.
+
+    Returns
+    -------
+    Callable to unregister the listener.
 
-    # Local variable to keep track of if the action has already been triggered
-    already_triggered = False
+    """
 
     @callback
-    def template_condition_listener(entity_id: str, from_s: State, to_s: State) -> None:
+    def state_changed_listener(
+        event: Optional[Event],
+        template: Template,
+        last_result: Optional[str],
+        result: Union[str, TemplateError],
+    ) -> None:
         """Check if condition is correct and run action."""
-        nonlocal already_triggered
-        template_result = condition.async_template(hass, template, variables)
-
-        # Check to see if template returns true
-        if template_result and not already_triggered:
-            already_triggered = True
-            hass.async_run_job(action, entity_id, from_s, to_s)
-        elif not template_result:
-            already_triggered = False
-
-    return async_track_state_change(
-        hass, template.extract_entities(variables), template_condition_listener
+        if isinstance(result, TemplateError):
+            _LOGGER.exception(result)
+            return
+
+        if event and last_result is not None:
+            if not _boolean_coerce(last_result) and _boolean_coerce(result):
+                hass.async_run_job(
+                    action,
+                    event.data.get("entity_id"),
+                    event.data.get("old_state"),
+                    event.data.get("new_state"),
+                )
+            return
+
+        if not _boolean_coerce(result):
+            return
+
+        #
+        # First run of the listener.
+        #
+        # Figure out an entity ID to
+        # pass back to the action because it expects one.
+        info = template.async_render_to_info(variables)
+        state = None
+        entity_id = None
+        for eid in info.entities:
+            state = hass.states.get(eid)
+            if state:
+                entity_id = eid
+                break
+        if not state:
+            for st in hass.states.async_all():
+                if info.filter_lifecycle(st.entity_id):
+                    state = st
+                    entity_id = st.entity_id
+                    break
+        hass.async_run_job(action, entity_id or MATCH_ALL, state, state)
```

Comment:
If the listener fires right away because it matches right away we pick the first entity that we find here. This seems like it is not going to be reliable.  I'm thinking we should get rid of this and send `MATCH_ALL` in this case since there isn't going to be an automation that fires for `entity_id` is on for 5 minutes when we just startup since we don't know that.

Reply:
I want to read back in the original discussion before I change this.

Reply:
Looks like the next change was to take out the initial firing per https://github.com/home-assistant/core/pull/23590#discussion_r304135152. That solves this